### PR TITLE
fix(spans): Do not overwrite `is_segment: true` if set by SDK

### PR DESCRIPTION
**Pull request by jjbayer**
We previously ignored `span.is_segment` if it was already set by the SDK, and set it to `false` for every span that has a `parent_id`. This is wrong because segments can have a `parent_id` in the trace.

### CODE REVIEW

**Comment by phacops** [comment]
This PR makes sense since a segment can have a `parent_span_id`. Trusting the `is_segment` flag from the SDK is the right move (or setting it if we don't see a `parent_span_id`).

**Review by phacops — APPROVED** [review]

**Review by iker-barriocanal — APPROVED** [review]

File: relay-server/src/services/processor/span/processing.rs [hunk 1]
```diff
@@ -414,12 +414,15 @@ fn get_normalize_span_config<'a>(
...
-    if is_segment {
+    // If `is_segment` was not set by the SDK and the span has no parent, assume it is a segment:
+    if is_root && span.is_segment.value().is_none() {
+        span.is_segment = true.into();
+    }
+
+    // If the span is a segment, always set to segment_id to the current span_id:
+    if span.is_segment.value() == Some(&true) {
         span.segment_id = span.span_id.clone();
     }
```
Comment (Dav1dde) [inline relay-server/src/services/processor/span/processing.rs:431]: Shouldn't we also check for this, if the SDK does that already we have to set `is_segment` to true as well.

Comment (jjbayer) [reply relay-server/src/services/processor/span/processing.rs:431]: You mean check the opposite direction? I.e. set `is_segment` if `segment_id == span_id`?

Comment (Dav1dde) [reply relay-server/src/services/processor/span/processing.rs:431]: Yes I think so, basically, expand `is_root` to:

```rs
let is_root = span.parent_span_id.is_empty() || span.segment_id == span.span_id;
```

Comment (jjbayer) [reply relay-server/src/services/processor/span/processing.rs:431]: Added this now. I also changed the logic to _not_ accept `is_segment: false` from the SDK if the condition above is met.

File: relay-server/src/services/processor/span/processing.rs [hunk 2]
```diff
@@ -414,12 +414,15 @@ fn get_normalize_span_config<'a>(
 fn set_segment_attributes(span: &mut Annotated<Span>) {
     let Some(span) = span.value_mut() else { return };
 
-    // TODO: A span might be a segment span even if the parent_id is not empty
-    // (parent within a trace). I.e. do not overwrite here.
-    let is_segment = span.parent_span_id.is_empty();
+    let is_definitely_segment = span.parent_span_id.is_empty() || span.segment_id == span.span_id;
```
Comment (Dav1dde) [inline relay-server/src/services/processor/span/processing.rs]: Hehe nice name

**Review by Dav1dde — APPROVED** [review]

File: relay-server/src/services/processor/span/processing.rs (hunk 2 above)
Comment (jjbayer) [reply relay-server/src/services/processor/span/processing.rs]: Hmm the second condition is also true when both are `None`... I'll add some more test coverage to not mess this up.

File: relay-server/src/services/processor/span/processing.rs [hunk 3]
```diff
@@ -413,13 +413,20 @@ fn get_normalize_span_config<'a>(
...
+        // The span is a segment if and only if the segment_id matches the span_id.
+        span.is_segment = (segment_id == span_id).into();
+    } else if span.parent_span_id.is_empty() {
+        // If the span has no parent, it is automatically a segment:
+        span.is_segment = true.into();
+    }
 
-    span.is_segment = Annotated::new(is_segment);
-    if is_segment {
+    // If the span is a segment, always set to segment_id to the current span_id:
```
Comment (jjbayer) [inline relay-server/src/services/processor/span/processing.rs]:
```suggestion
    // If the span is a segment, always set the segment_id to the current span_id:
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,7 @@
 
 - Fix performance regression in disk spooling by using page counts to estimate the spool size. ([#3379](https://github.com/getsentry/relay/pull/3379))
 - Perform clock drift normalization only when `sent_at` is set in the `Envelope` headers. ([#3405](https://github.com/getsentry/relay/pull/3405))
+- Do not overwrite `span.is_segment` if already set by SDK. ([#3411](https://github.com/getsentry/relay/pull/3411))
```
Comment (jjbayer) [inline CHANGELOG.md]:
```suggestion
- Do not overwrite `span.is_segment: true` if already set by SDK. ([#3411](https://github.com/getsentry/relay/pull/3411))
```

File: tests/integration/test_spans.py [hunk 5]
```diff
@@ -308,8 +308,8 @@ def envelope_with_spans(
                     {
                         "description": "https://example.com/p/blah.js",
                         "op": "resource.script",
-                        "span_id": "bd429c44b67a3eb1",
-                        "segment_id": "968cff94913ebb07",
+                        "span_id": "b0429c44b67a3eb1",
+                        "segment_id": "b0429c44b67a3eb1",
```
Comment (jjbayer) [inline tests/integration/test_spans.py:312]: I gave this span a new span_id and the same segment_id, to make sure it is classified as `is_segment`.

File: tests/integration/test_spans.py [hunk 6]
```diff
@@ -526,19 +526,19 @@ def test_span_ingestion(
...
+            "span_id": "b0429c44b67a3eb1",
             "start_timestamp_ms": int(start.timestamp() * 1e3),
             "trace_id": "ff62a8b040f340bda5d830223def1d81",
         },
         {
             "description": r"test \" with \" escaped \" chars",
             "duration_ms": 1500,
             "exclusive_time_ms": 345.0,
-            "is_segment": True,
+            "is_segment": False,
```
Comment (jjbayer) [inline tests/integration/test_spans.py:537]: This span has a different segment_id than span_id, so it's not a segment.

File: tests/integration/test_spans.py [hunk 7]
```diff
@@ -598,7 +598,7 @@ def test_span_ingestion(
 
     # If transaction extraction is enabled, expect transactions:
     if extract_transaction:
-        expected_transactions = 5
+        expected_transactions = 3
```
Comment (jjbayer) [inline tests/integration/test_spans.py:601]: Two spans that were previously mis-classified as segment are now non-segments.

File: tests/integration/test_spans.py [hunk 8]
```diff
@@ -1230,8 +1225,8 @@ def test_span_ingestion_with_performance_scores(
                         },
                         "profile_id": "3d9428087fda4ba0936788b70a7587d0",
                         "op": "ui.interaction.click",
-                        "span_id": "bd429c44b67a3eb1",
-                        "segment_id": "968cff94913ebb07",
+                        "span_id": "cd429c44b67a3eb1",
+                        "segment_id": "cd429c44b67a3eb1",
```
Comment (jjbayer) [inline tests/integration/test_spans.py:1229]: I checked some real `ui.interaction.click` spans in the product and they all have `span_id == segment_id`.

File: tests/integration/test_spans.py (hunk 6 above)
Comment (Dav1dde) [reply tests/integration/test_spans.py:537]: Can you put this as a comment in the code?

File: tests/integration/test_spans.py (hunk 5 above)
Comment (Dav1dde) [reply tests/integration/test_spans.py:312]: Also please put that as a comment in the code

Comment (Dav1dde) [reply tests/integration/test_spans.py:312]: Ah already merged, dw about it. I'll throw up a PR.

Comment (Dav1dde) [reply tests/integration/test_spans.py:312]: #3453

**Pull request merged**